### PR TITLE
feat(middleware): create a generic middleware to collect method metrics.

### PR DESCRIPTION
**Pull request by andysim3d**
[Closes/Fixes] #787

## Proposed Changes

  -
  -
  -

### CODE REVIEW

File: crates/task/src/metriclayer.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    fn call(&mut self, request: http::Request<Body>) -> Self::Future {
+        let uri = request.uri().clone();
+        let method_name = uri.path().split('/').last().unwrap_or("unknown");
+        {
+            let mut metric = method_metric.lock().unwarp();
+            MethodMetrics::increment_num_requests(self.service_name.as_str(), method_name);
+            MethodMetrics::increment_open_requests(self.service_name.as_str(), method_name);
+        }
+        ResponseFuture{
```
Comment (andysim3d) [inline crates/task/src/metriclayer.rs:91]: consider not redefine customize Future.

File: crates/task/src/metriclayer.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    start_time: Instant,
+    method_name: String,
+    service_name: String,
+}
+
+impl<S> Service<RequestInfo> for MetricsMiddleware<S>
+where
+    S: Service<RequestInfo>,
+{
+    type Response = S::Response;
```
Comment (andysim3d) [inline crates/task/src/metriclayer.rs:75]: add a method_extractor here, and make that generic.

File: crates/rpc/src/task.rs [hunk 3]
```diff
@@ -186,12 +185,11 @@ where
             .layer(ProxyGetRequestLayer::new("/health", "system_health")?)
             .timeout(self.args.rpc_timeout);
 
-        let rpc_middleware =
-            RpcServiceBuilder::new().layer(RpcMetricsMiddlewareLayer::new(&module));
+        let rpc_metric_middleware = MetricsLayer::new("rpc".to_string(), "rpc".to_string());
```
Comment (dancoombs) [inline crates/rpc/src/task.rs:188]: Lets add "server" here somewhere so we know these are requests we're serving, not the client.

File: crates/task/src/grpc/metrics.rs [hunk 4]
```diff
@@ -13,145 +13,151 @@
...
+//     // Increment the number of gRPC errors for a given method and service.
+//     fn increment_rpc_error_count(method_name: &str, scope: &str) {
+//         metrics::counter!("grpc_error_count", "method_name" => method_name.to_string(), "service" => scope.to_string()).increment(1)
+//     }
+
+//     // Record the latency of a request for a given method and service.
+//     fn record_request_latency(method_name: &str, scope: &str, latency: Duration) {
+//         metrics::histogram!("grpc_request_latency", "method_name" => method_name.to_string(), "service" => scope.to_string()).record(latency)
+//     }
+// }
```
Comment (dancoombs) [inline crates/task/src/grpc/metrics.rs:163]: Any reason not to delete?

File: crates/task/src/metriclayer.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
+// This file is part of Rundler.
```
Comment (dancoombs) [inline crates/task/src/metriclayer.rs:1]: For the file name, lets either use one word `metrics.rs` (preferred if its descriptive enough) or snake case `metrics_layer.rs`

File: crates/task/src/metriclayer.rs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+use std::{
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+use futures::{future::BoxFuture, FutureExt};
+use tower::{Layer, Service};
+
+/// Trait to expose request method name.
+pub trait RequestInfo {
```
Comment (dancoombs) [inline crates/task/src/metriclayer.rs:25]: Can we make this name more descriptive of its functionality? `RequestInfo` seems misleading

File: crates/task/src/grpc/metrics.rs (hunk 4 above)
Comment (andysim3d) [reply crates/task/src/grpc/metrics.rs:163]: vs-code fools me - it didn't save the file after I deleted them.

File: crates/task/src/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+use std::{
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+use futures::{future::BoxFuture, FutureExt};
+use tower::{Layer, Service};
+
+/// Trait to expose request method name.
+pub trait RequestMethoedNameInfo {
```
Comment (dancoombs) [inline crates/task/src/metrics.rs:25]: Typo: `RequestMethoedNameInfo` -> `RequestMethodNameInfo`

File: crates/rpc/src/rpc_metrics.rs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+// If not, see https://www.gnu.org/licenses/.
+
+use rundler_task::metrics::RequestMethoedNameInfo;
+use jsonrpsee::types::Request;
+
+impl RequestMethoedNameInfo for Request {
+    fn get_method_name(&self) -> String {
+        self.method_name().to_string()
+    }
+}
```
Comment (dancoombs) [inline crates/rpc/src/rpc_metrics.rs:21]: nit: newline at end of file, I thought our formatter added these by default?

**Review by dancoombs — COMMENTED** [review]
You should also be able to implement this for Alloy in the `provider` crate

**Comment by andysim3d** [comment]
> You should also be able to implement this for Alloy in the `provider` crate

yup I will do that in other pr. git won't allow me to split commits into multiple prs.

**Comment by dancoombs** [comment]
> > You should also be able to implement this for Alloy in the `provider` crate
> 
> yup I will do that in other pr. git won't allow me to split commits into multiple prs.

What do you mean? You should be able to add it to this PR?

**Comment by andysim3d** [comment]
> > > You should also be able to implement this for Alloy in the `provider` crate
> > 
> > 
> > yup I will do that in other pr. git won't allow me to split commits into multiple prs.
> 
> What do you mean? You should be able to add it to this PR?

I can - but prefer small prs.

File: crates/provider/src/traits/metrics.rs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+impl RequestExtractor<RequestPacket> for RPCMethodExtractor {
+    fn get_method_name(req: &RequestPacket) -> String {
+        match req {
+            RequestPacket::Single(request) => {
+                request.method().to_string()
+            }
+            _ => {
+                // can't extract method name for batch.
+                "unknown".to_string()
```
Comment (dancoombs) [inline crates/provider/src/traits/metrics.rs:29]: Lets call this `"batch"`

File: crates/rpc/src/rpc_metrics.rs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along with Rundler.
+// If not, see https://www.gnu.org/licenses/.
+
+use jsonrpsee::types::Request;
+use rundler_types::task::traits::RequestExtractor;
+
+#[derive(Copy)]
```
Comment (dancoombs) [inline crates/rpc/src/rpc_metrics.rs:17]: You likely need to derive Clone here as well

File: crates/task/src/metrics.rs [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+            inner: inner,
+            service_name: service_name.clone(),
+            protocal: protocal,
+            _request_extractor_: PhantomData,
+            _request_type_: PhantomData,
+        }
+    }
+}
+
+impl<S, T, Request> Service<Request> for MetricsMiddleware<S, T, Request>
```
Comment (dancoombs) [inline crates/task/src/metrics.rs:85]: A feature that I'm realizing now that this is missing is related to response codes. Our existing ethers based metrics layer would extract those:

https://github.com/alchemyplatform/rundler/blob/2a0e8be0137579bf6d1c4c13f0609f738adc7965/crates/provider/src/ethers/metrics_middleware.rs#L86

Lets do this in a follow up, but we may need to also add a `ResponseExtractor` that can get the code out and then have that be part of the metrics.

Comment (dancoombs) [reply crates/task/src/metrics.rs:85]: Mind creating a follow up ticket and assigning it to milestone v0.4 and yourself?

**Comment by dancoombs** [comment]
Few last comments, but other than that LGTM! Will ✅ after the small updates.

File: crates/task/src/metrics.rs (hunk 11 above)
Comment (andysim3d) [reply crates/task/src/metrics.rs:85]: created #812

**Review by dancoombs — APPROVED** [review]
Ship it 🚢

**Pull request merged**